changed buttons to links in assessment dashboard and teams page

julian-christl opened the pull request:
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, editor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [X] Client: I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client/).
- [X] Client: I added multiple integration tests (Jest) related to the features (with a high test coverage)
- [X] Client: I documented the TypeScript code using JSDoc style.
- [X] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language]

 **Testing:**
- [x] Documentation
- [x] Tests (1st person)
- [x] Tests (2nd person)
- [x] Source Code (1st person)
- [x] Source Code (2nd person)
- [x] Team-Page
- [x] Course Assessment Dashboard
- [x] Exam Assessment Dashboard (Test Run)
- [x] Exam Assessment Dashboard (Default)

### Motivation and Context
**Team-Page:** When assessing team exercises, you often need to open multiple old exercises aside to compare everything for the current assessment. Currently, you have to open the same page in multiple tabs and navigate to each exercise instead of just quickly opening all pages in a new tab.
**Assessment-Dashboard:** Sometimes, you need to compare different assessments or complaints, so it's the same here. However, this was just another button occurrence I found, so I updated it as well.

### Description
I replaced some buttons which opened assessment editors with links.
=> Previously, one could only open them with left-click, now left-click, context-menu and middle-click are all available for these buttons.
Technically this relates to the user interface. However, nothing was visually changed, so I didn't add any screenshots.
I furthermore removed the now unreferenced methods and their tests and added tests for my newly created methods.
I also added a short documentation entry.

## Steps for Testing
**Documentation**
Please check whether it's consistent with the rest of the documentation and makes sense.

### Code Review
Please check this thoroughly since I'm new to Artemis. Let me know if I should improve something.
I suggest that you start with the TypeScript changes before looking at the HTML. Also, you might want to review component by component.
After that, you can check out the tests. I'd like to have your opinion on whether they are too dumb to miss some test cases. With new systems, I always need some time to adapt my testing skills to that environment.

### Teams (not changed from the first time)
**Requirements:**
- Instructor Account and additional account for the submission if needed
- Team exercise with a team (containing instructor or the mentioned additional one) and no submission of this team
1. Navigate to the Exercises of your course
2. Open the Teams section of your team exercise
3. Open your team view by clicking on the short name
4. Make sure that you can't assess the exercise (disabled link)
5. Submit a submission for this team
6. Make sure that the previous disabled button is navigatable like a link (context menu, middle mouse click to open in a new tab, etc.)
7. Make sure that after opening this the first time, actions from Step 6 are still possible. Left-click should be still an option.

### Assessment Dashboard (changed)
**Requirements**
- Instructor Account and additional account for the submission if needed
- An course exercise with two submissions
1. Navigate to the assessment dashboard
2. Open the dashboard for the corresponding exercise
3. Make sure the assessment button behaves as in 6./7. above when starting an assessment
4. Make sure it behaves the same after opening it the first time and after **assessing the submission**
5. Assess the second exercise
5. File a complaint about the assessment from step 4
6. Make sure that the button for the complaint behaves the same (+ after the complaint was resolved)
7. File a feedback request about the assessment from step 5
8. Make sure that the button for the feedback request behaves the same (+ after the request was resolved)

### Exam Assessment Dashboard Test Run (not changed)
**Requirement**
- Instructor Account
- Exam with at least one exercise
1. Conduct a test run. Assess your test run and make sure that when clicking the button in the assessment dashboard, the query parameters include the test run key
2. Test the same way as you did in the course assessment dashboard minus the feedback request

### Exam Assessment Dashboard (not changed)
**Requirement**
- Instructor Account
- Exam with at least one exercise, conducted once
1. Test the same way as you did in the previous sections. Make sure that the query parameters do not contain a test run

## Screenshots
Teams:
![ee691e0a63a19ddee330e842bb722f91](https://user-images.githubusercontent.com/64264938/129274886-32361a55-da42-4c8a-911e-0560bd687f6d.png)
Assessment Dashbards (Here the course AD, but the other two look the same minus More Feedback Request)
![25279a421e009dbf8641b092f0eecd3b](https://user-images.githubusercontent.com/64264938/129274889-77f04072-43af-43b8-a4e9-1d4a7dc58acf.png)

## Test Coverage
exercise-assessment-dashboard.component.ts 83.8%
team-participation-table.component.ts 93.24%

julian-christl commented:
## Important
If anyone knows about additional buttons that fall in this category, let me know. I either add them to this PR as well or do a follow-up.

FuchsDominik reviewed (changes requested):
The "open assessment" button does not behave as a link
![Screenshot 2021-08-10 at 17 12 35](https://user-images.githubusercontent.com/44401560/128892740-87672321-e938-4dfa-9646-6ef11b667acf.png)

julian-christl commented:
I adapted the PR description accordingly, added what parts changed and what parts didn't.
I also added some checkboxes for testing so please write down what you tested so I can check them to see what was done and what wasn't (I can also change the checkboxes if necessary so don't worry :D)

rs-maker commented on docs/dev/guidelines/client.rst:
```diff
@@ -63,7 +63,14 @@ Some general aspects:
+8. Buttons and Links
+========
+
+1. Be aware that Buttons navigate only in the same tab while Links provide the option to use the context menu or a middle-click to open the page in a new tab. Therefore:
+2. Buttons are best used to trigger certain functionalities
+3. Links are best for navigating on Artemis
+
```
I think this is great but if you look at the other Documentation points they include examples. So maybe show an example like:
Dont do: <button (click)="openSomeOtherPage()"> ... <\/button>
Do: <a [routerLink]="getLinkToSomeOtherPage()"
            [queryParams]="getQueryParamsForPage()"> ... <\/a>

julian-christl replied:
done

rs-maker reviewed (changes requested):
The code looks good but you have some client style issues. Running yarn lint:fix should fix those.
I also left a comment on your doc entry. 
Edit: i only did code review since the functionality was reviewed yesterday in the testing session.

julian-christl commented:
@rs-maker Thanks for reviewing. Keep in mind though that some functionality changed. So please let me know what you tested yesterday and if you have the time you can test the rest (Assessment Dashboard).

Malyuk-A reviewed (changes requested):
Code looks good 👍
The only thing I would currently change is to remove the unused/commented parts in the test class. (e.g. import Complaint & describe 'openAssessmentEditor' line 565-616) But because you just commented it in yesterday (11.08) I suppose this is WIP?

Malyuk-A reviewed (commented):
I tested your PR on TS3.
I testest the Teams, Assessment Dashboard, and Exam Assessment Dashboard.
It seems to work fine, even though I had difficulties understanding the expected behavior.
E.g. What do you mean at Team 6? Do you mean the Submit button should behave like a link? If yes why?
![image](https://user-images.githubusercontent.com/65814168/129227852-141f6000-9754-4e0d-b2b0-a97260075555.png)

What also confuses me a bit is the "behaves the same" do you mean it behaves like a button or a link?

What might help me and others (is not a must but maybe it can be a useful trick)
to use colors and brackets e.g. : 
Step A. [🔵 like button]
Step B. [🟢 like link]
 or simply
Step A. 🔵
Step B. 🟢

julian-christl commented:
@Malyuk-A Thanks for testing. I added two screenshots that mark the mentioned buttons.
The only behaviour I expect is that of a link => context menu, middle click, left click
Not other behaviour intended for the mentioned button look-alike links (a bit difficult to differentiate that here since they technically look like buttons :D)

And yes, I tried to structure it to the best of my abilities but apparently, that is not enough 😅 
However, in this case, that would help much as you hopefully can see yourself since there is only one expected behaviour for all.

julian-christl commented:
@Malyuk-A And thanks for pointing the unused code out. Not sure how I missed that when committing but yeah, fixed. :D

Malyuk-A reviewed (approved):
Tested again on TS3. Works as described. 👍
Great enhancement.
Code looks good (since my first review the unused code has been removed)

rs-maker reviewed (approved):
tested it on ts2 works great. code looks good too :+1:

MDK5 commented on docs/dev/guidelines/client.rst:
```diff
@@ -63,7 +63,14 @@ Some general aspects:
+
+1. Be aware that Buttons navigate only in the same tab while Links provide the option to use the context menu or a middle-click to open the page in a new tab. Therefore:
+2. Buttons are best used to trigger certain functionalities (e.g. ``<button (click)='deleteExercise(exercise)'>...</button``)
+3. Links are best for navigating on Artemis (e.g. ``<a [routerLink]='openExerciseEditor(exercise)' [queryParams]='getQueryParamsForEditor(exercise)'>...</a>``)
```
```suggestion
3. Links are best for navigating on Artemis (e.g. ``<a [routerLink]='getLinkForExerciseEditor(exercise)' [queryParams]='getQueryParamsForEditor(exercise)'>...</a>``)
```

MDK5 commented on src/main/webapp/app/exercises/shared/dashboards/tutor/exercise-assessment-dashboard.component.ts:
```diff
@@ -598,47 +598,63 @@ export class ExerciseAssessmentDashboardComponent implements OnInit {
             if (!submissionToView.results) {
                 submissionToView.results = [];
             }
             submissionToView.results = submissionToView.results?.filter((result) => result.assessmentType !== AssessmentType.AUTOMATIC);
```
This line and the if above could be shortened to 
```suggestion
            submissionToView.results = submissionToView.results?.filter((result) => result.assessmentType !== AssessmentType.AUTOMATIC) ?? [];
```
however I am not sure if it improves readability

jpbernius replied:
I guess this is ok. Otherwise, we can remove the `?` and put the `filter` in an `else` block.

julian-christl replied:
I decided to leave it as a standard if clause but I changed it to an if-else clause as suggested because I think this is more intuitive to people not as familiar with TS syntax.

MDK5 commented on src/main/webapp/app/exercises/shared/dashboards/tutor/exercise-assessment-dashboard.component.ts:
```diff
@@ -598,47 +598,63 @@ export class ExerciseAssessmentDashboardComponent implements OnInit {
+     * This complaint has to be a true complaint or else issues can arise.
+     * @param submission
+     */
+    getSubmissionToViewFromComplaintSubmission(submission: Submission) {
```
```suggestion
    getSubmissionToViewFromComplaintSubmission(submission: Submission): Submission | undefined {
```
I'd prefer writing out the return type

MDK5 reviewed (approved):
Tested on ts1 with almost all buttons, all behaved like links.
Also reviewed code, changes look good to me, just found a few nitpicks.

jpbernius commented on src/main/webapp/app/exercises/shared/dashboards/tutor/exercise-assessment-dashboard.component.ts:
```diff
@@ -598,47 +598,63 @@ export class ExerciseAssessmentDashboardComponent implements OnInit {
+     * @param submission
+     */
+    getSubmissionToViewFromComplaintSubmission(submission: Submission) {
         const submissionToView = this.submissionsWithComplaints.filter((dto) => dto.submission.id === submission.id).pop()?.submission;
```
Is this the same thing?
```suggestion
        const submissionToView = this.submissionsWithComplaints.find((dto) => dto.submission.id === submission.id)?.submission;
```

jpbernius reviewed (changes requested):
Looking good. Can we integrate the suggestions from @MDK5 before merging this?

argertboja reviewed (approved):
Tested on TS1 and everything works fine. Code changes look good as well! 👍🏼

MDK5 reviewed (approved):
Re-Approve, comments got adressed
